### PR TITLE
Follow up on #12327: circuit construction in Rust

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

#12327 ported the utility functions for `synth_permutation_basic` to Rust, but not the circuit construction which, at the time, wasn't possible. Now that we can construct circuits in Rust directly, this PR follows up by doing exactly that (and also adds a reno).

### CODE REVIEW

File: crates/accelerate/src/permutation.rs [hunk 1]
```diff
@@ -111,10 +116,31 @@ fn _get_ordered_swap(py: Python, permutation_in: PyArrayLike1<i64>) -> PyResult<
...
+        get_ordered_swap(&view).iter().map(|(i, j)| {
+            (
+                StandardGate::SwapGate,
+                smallvec![],
+                smallvec![Qubit(*i as u32), Qubit(*j as u32)],
+            )
+        }),
+        Param::Float(0.0),
+    )
+    .expect("Something went wrong in Qiskit's Python realm"))
```
Comment (mtreinish) [inline crates/accelerate/src/permutation.rs]: I'm going to enjoy when we have a user complaining about an uncatchable `PanicException` complaining about the python realm when calling `Qubit()` fails unexpectedly. :p

File: crates/accelerate/src/permutation.rs [hunk 2]
```diff
@@ -111,10 +116,31 @@ fn _get_ordered_swap(py: Python, permutation_in: PyArrayLike1<i64>) -> PyResult<
...
 
+#[pyfunction]
+#[pyo3(signature = (pattern))]
+fn _synth_permutation_basic(py: Python, pattern: PyArrayLike1<i64>) -> PyResult<CircuitData> {
+    let view = pattern.as_array();
+    let num_qubits = view.len();
+    Ok(CircuitData::from_standard_gates(
+        py,
+        num_qubits as u32,
+        get_ordered_swap(&view).iter().map(|(i, j)| {
```
Comment (mtreinish) [inline crates/accelerate/src/permutation.rs]: Is there a reason to leave this as a `pyfunction`? It seems more natural to have this take an `ArrayView1` or a slice and leave it as solely a private rust function.

File: crates/accelerate/src/permutation.rs (hunk 1 above)
Comment (Cryoris) [reply crates/accelerate/src/permutation.rs]: Yup that'll inevitably happen at some point 😅 is there a better way of handling it right now? I was mimicking how it's done in other place so far 😬

File: crates/accelerate/src/permutation.rs (hunk 2 above)
Comment (Cryoris) [reply crates/accelerate/src/permutation.rs]: It's currently also used in the construction of the `PermutationGate` (and circuit, they are doubled atm). But I can change that code to instead use `synth_permutation_basic` which would be a bit better as it is actually part of the API as opposed to `_get_ordered_swap`.

Comment (Cryoris) [reply crates/accelerate/src/permutation.rs]: Done in b0fcf2a

File: crates/accelerate/src/permutation.rs [hunk 3]
```diff
@@ -95,26 +109,30 @@ fn _inverse_pattern(py: Python, pattern: PyArrayLike1<i64>) -> PyResult<PyObject
...
+            (
+                StandardGate::SwapGate,
+                smallvec![],
+                smallvec![Qubit(*i as u32), Qubit(*j as u32)],
+            )
+        }),
+        Param::Float(0.0),
+    )
+    .expect("Something went wrong in Qiskit's Python realm"))
```
Comment (mtreinish) [inline crates/accelerate/src/permutation.rs]: Based on our offline discussion I think:

```suggestion
    CircuitData::from_standard_gates(
        py,
        num_qubits as u32,
        get_ordered_swap(&view).iter().map(|(i, j)| {
            (
                StandardGate::SwapGate,
                smallvec![],
                smallvec![Qubit(*i as u32), Qubit(*j as u32)],
            )
        }),
        Param::Float(0.0),
    )
```
should work and answer: https://github.com/Qiskit/qiskit/pull/12605/files#r1647033844

Comment (mtreinish) [reply crates/accelerate/src/permutation.rs]: Of course this will fail rustfmt and I also didn't test it locally, but it's the basic idea.

Comment (Cryoris) [reply crates/accelerate/src/permutation.rs]: Thanks! Done in b860eb0 🙂

**Comment by Cryoris** [comment]
I've also added the module structure in bb03417 as discussed with @ShellyGarion, so after this merged we can follow it in the other open PRs 🙂

**Review by alexanderivrii — APPROVED** [review]
LGTM! Special thanks for organizing the synthesis directory structure.

**Review by ShellyGarion — APPROVED** [review]

**Pull request merged**